### PR TITLE
app: add syntax highlighting support via Chroma

### PR DESCRIPTION
**Pull request by slimsag**
Today, Sourcegraph App has no syntax highlighting. This is for the reasons outlined in a new comment:

https://github.com/sourcegraph/sourcegraph/blob/96bcec929f3569c9e87755868e5458677eee8deb/cmd/frontend/internal/highlight/highlight.go#L452-L460

In this PR I pull in [Chroma](https://github.com/alecthomas/chroma), the most popular Go syntax highlighter used by e.g. Hugo. It is effectively a Go port of Pygments, and seems to have good compatibility with Pygments grammars and decent performance. It supports [a decent chunk of languages](https://github.com/alecthomas/chroma#supported-languages).

One challenge is that Chroma TokenTypes and SCIP SyntaxKinds do not exactly agree / match 1:1. In some cases, Chroma is more detailed, while in others SCIP is more detailed. In some cases Chroma has strange choices, like e.g. when highlighting Go code "foo" in "package foo" is considered a builtin. I did a best-effort mapping, and for most languages it seems to work pretty well.

I don't want to pretend this is a perfect solution (is there ever such a thing with syntax highlighting?); it's certainly not, and syntect is definitely better - but I think this is *good enough* for now for App, and it brings us from zero -> colors which is a clear win.

## Test plan

This change only affects App.

To test I manually reviewed a few handfuls of files:

Decent:

* Makefile https://user-images.githubusercontent.com/3173176/218580862-25d9a9a2-fe02-404e-b90d-0172409286f1.png
* Go https://user-images.githubusercontent.com/3173176/218580943-8b646a2f-8f65-49fd-bbd1-f9052e5239a9.png
* Typescript https://user-images.githubusercontent.com/3173176/218582523-fbd60a5f-d43e-4aa3-bd85-2a4efce0b7bd.png
* Bazel https://user-images.githubusercontent.com/3173176/218582689-d64cd875-3a46-4aa1-bf02-07c949cdef9b.png
* Nix https://user-images.githubusercontent.com/3173176/218582776-a1cbb10e-f71f-4044-9999-b0d144f2f25d.png
* Python https://user-images.githubusercontent.com/3173176/218582931-131f8737-87f0-463c-a121-117cd23667db.png
* C# https://user-images.githubusercontent.com/3173176/218583003-6a8a5d55-3115-4c6d-a269-099407322378.png
* SCSS https://user-images.githubusercontent.com/3173176/218583049-13010314-a139-4a7f-af6b-a1c8d26166c7.png
* Scala https://user-images.githubusercontent.com/3173176/218583104-9d1877db-c9be-49b5-bb9f-6b8cea37b9f1.png
* C https://user-images.githubusercontent.com/3173176/218583146-0fbf73f9-fd0a-4adf-8398-c8c4c2a902d9.png
* SQL https://user-images.githubusercontent.com/3173176/218583191-555f9c9a-ae58-4ca8-a752-4543ab9bc460.png
* CSS https://user-images.githubusercontent.com/3173176/218583242-c6c18357-01b4-4975-b23f-d19c04d565d7.png
* YAML https://user-images.githubusercontent.com/3173176/218581046-e106444e-4626-4789-a967-de144f4c0199.png
* Dockerfile https://user-images.githubusercontent.com/3173176/218581198-746d575f-ff84-45af-93c7-ee11150ea38d.png
* Plain text https://user-images.githubusercontent.com/3173176/218581291-f40fef95-8694-48ae-b41f-78c05287f32d.png
* ps1 https://user-images.githubusercontent.com/3173176/218581520-2ade3d56-5915-4b31-a202-86831941fcfb.png
* Makefile https://user-images.githubusercontent.com/3173176/218581779-e64bce10-8c37-425d-a5e4-fdded98b9f1a.png
* Shell https://user-images.githubusercontent.com/3173176/218581911-8a0d9753-4f97-47c5-a2e0-dde2576e9968.png

Poor:

* JSON https://user-images.githubusercontent.com/3173176/218581465-2894fa9c-7ca1-4336-855c-cf602a22d343.png
* Some YAML files (but not all)?
* go.mod https://user-images.githubusercontent.com/3173176/218582069-838f7ce1-6736-4a4b-aec8-d21dbe339596.png

No highlighting:

* Markdown
* Procfile
* .patch

### CODE REVIEW

**Comment by olafurpg** [comment]
I strongly encourage us to figure out how to make syntax-highlighter work with Sourcegraph App instead of introducing a third highlighting solution. We're increasingly relying on SCIP metadata from syntax-highlighter to power various functionality so I expect the feature gap between App and our standard product to only increase if we use Chroma.

Not a blocking comment to merge this PR, just voicing concerns about going this route.

**Comment by slimsag** [comment]
@olafurpg yeah that makes sense, I want to be clear this is just a temporary solution until we can validate that App is actually worth continuing to invest in - does that make sense?

**Comment by slimsag** [comment]
The way I am thinking about this:

* If App goes well at Starship MVP and sees actual usage, we can invest in highlighting (investigate how to remove http-server-stabilizer from syntect-server, investigate how to use it as a library via CGO + integrate Rust into our cross-compilation toolchain)
* If App doesn't go well at Starship, we at least didn't waste too much time thinking about highlighting / just chose the quick option to get colors on the screen. We can easily rip the code out later since it's just an `if` branch.

**Comment by olafurpg** [comment]
@slimsag Sounds good to me. SCIP is designed exactly to support this kind of situation where we want to plug in a new syntax highlighting backend. Our previous blob view inlined raw HTML from syntect and I suspect it would have been a lot more complicated to replace add Chroma. I'm thrilled that this diff is only 140 loc!

**Comment by slimsag** [comment]
Definitely, I like what you've done with SCIP highlighting overall! It did make this change quite easy to integrate

File: cmd/frontend/internal/highlight/chroma.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+		return nil, errors.Wrap(err, "Tokenise")
+	}
+
+	formatter := &chromaSCIPFormatter{}
+	occurrences, err := formatter.Format(iterator)
+	if err != nil {
+		return nil, errors.Wrap(err, "Format")
+	}
+	return &scip.Document{
+		Language:     scip.Language_Go.String(),
```
Comment (eseliger) [inline cmd/frontend/internal/highlight/chroma.go:37]: Should this be static?

**Review by eseliger — APPROVED** [review]
Nice non-intrusive implementation 😊

File: cmd/frontend/internal/highlight/chroma.go (hunk 1 above)
Comment (olafurpg) [reply cmd/frontend/internal/highlight/chroma.go:37]: The `Language` field can be empty because it's not used by the client.

**Review by olafurpg — COMMENTED** [review]
FYI, we extensively use snapshot testing for SCIP payloads. Our Go bindings provide a `FormatSnapshot` helper that you can use. However, it currently doesn't print `SyntaxKind` right now but I can quickly approve a PR adding the change. 

 https://sourcegraph.com/github.com/sourcegraph/scip@3050326/-/blob/bindings/go/scip/testutil/format.go

We have syntect highlighting tests here 

https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/tree/docker-images/syntax-highlighter/crates/sg-syntax/src/snapshots/syntect_files

I'm working on a PR adding tree-sitter highlighting for more languages and the PR adds a lot of higher quality snapshot tests in the sibling "files/" directory (the directory should be named "tree-sitter-files/" or something like that)

**Comment by olafurpg** [comment]
Also FYI, our blob view has a concept of "interactive tokens", which are tokens that are likely to have navigation data and should therefore load a popover even if there's no definition data. The set of interactive `SyntaxKind` is defined here https://sourcegraph.com/github.com/sourcegraph/sourcegraph@4e4ff62a3d7a60abb0616aba500d7b24508d672d/-/blob/client/web/src/repo/blob/codemirror/occurrence-utils.ts?L13

**Comment by slimsag** [comment]
Ready to merge but will wait for code freeze to end.

Also added a TODO for myself on our board to follow up on snapshot testing here, that will be quite helpful

**Comment by slimsag** [comment]
I am told code freeze is over and we can merge things again. This is also a low-risk change (only affects app) so I am merging.

**Pull request merged**